### PR TITLE
Cadence HiFi Mini NN Library: Integrated optimized kernels

### PR DESCRIPTION
**Pull request by bhanuprakashbv**
Integrated Cadence HiFi Mini optimized functions for 
fully_connected, softmax and svdf kernels.
These optimized functions gives better performance for keyword_benchmark application.

### CODE REVIEW

File: tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc [hunk 1]
```diff
@@ -26,6 +47,7 @@ limitations under the License.
 #include "tensorflow/lite/kernels/kernel_util.h"
 #include "tensorflow/lite/micro/kernels/xtensa_hifimini/fixedpoint_utils.h"
 
+#include "xtensa_tf_micro_common.h"
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc:50]: Here and below, this should be a full path "tensorflow/lite/kernels/xtensa_hifimini/xtensa_tf_micro_common"

File: tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc [hunk 2]
```diff
@@ -202,11 +229,32 @@ TfLiteStatus EvalQuantizedInt8(TfLiteContext* context, TfLiteNode* node,
...
-      GetTensorShape(bias), GetTensorData<int32_t>(bias),
-      GetTensorShape(output), GetTensorData<int8_t>(output));
+  {
+    int ret, b, weight_depth, out_depth, batches;
+    int8_t * p_out = GetTensorData<int8_t>(output);
+    weight_depth = GetTensorShape(filter).Dims(GetTensorShape(filter).DimensionsCount()-1);
+    out_depth = GetTensorShape(output).Dims(GetTensorShape(output).DimensionsCount()-1);
+    batches = FlatSizeSkipDim(GetTensorShape(output), GetTensorShape(output).DimensionsCount()-1);
+
+    //TODO: Use xa_nn_fully_connected_sym8xasym8s_asym8s? the kernel tests fail with it.
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc:239]: This is probably a bug in the kernel tests.  I will have a a look internally at why we are testing against asymmetric weights for fully_connected_test.

File: tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc [hunk 3]
```diff
@@ -91,168 +114,89 @@ void EvalIntegerSVDF(TfLiteContext* context, TfLiteNode* node,
   int16_t* const state_ptr = GetTensorData<int16_t>(activation_state_tensor);
 
   // Left shift the activation_state.
+  if(((int)state_ptr&0x3) == 0)
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc:117]: avoid c-style cast + split out this logic with a comment for clarify.

File: tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc [hunk 4]
```diff
@@ -23,13 +44,19 @@ limitations under the License.
...
 namespace micro {
 namespace activations {
 namespace {
 
 struct OpData {
+  int32_t input_multiplier;
+  int32_t input_left_shift;
+  int32_t diff_min;
+  int scratch_tensor_index;
   uint16_t* exp_lut;
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc:60]: If exp_lut is unused, can we remove it?

File: tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc [hunk 5]
```diff
@@ -1,3 +1,24 @@
...
+  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+  * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+  * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+  * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+  * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+  ******************************************************************************/
+
 /* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc:22]: General question about softmax - have you compared the lookup table based version with your optimized version?  I found that calculating exp(-x) is quite expensive in general, and since we have 8-bit inputs we can save a lot of cycles by e(-beta * scale * x) for all 256 possible inputs in a LUT populated in prepare().  If the LUT is faster, these optimizations would be awesome to have in prepare where the LUT is populated.  If the your optimized version is close in cycle count then it may not be worth the effort to convert to LUT based solution.

**Review by njeffrie — COMMENTED** [review]
These are awesome changes - thanks for all the hard work!  These changes will require quite extensive internal testing and validation as well as some preparation to roll forward, so this may end up approved but not submitted for a while we do that validation and preparation.

**Comment by gbaned** [comment]
@bhanuprakashbv Thank you for your contribution. Can you please sign CLA? Thanks!

**Comment by bhanuprakashbv** [comment]
Hi,

Could you please help me to sign CLA. I did already, which is showing like below.
I don’t know how to add my contributions without CLA issues.

Thanks,
Bhanu
[cid:image002.jpg@01D62E83.185E8750]

**Comment by bhanuprakashbv** [comment]
@googlebot I signed it!
After signing CLA, status is like below.
I don’t know whether it is completed or anything else do I need to do.
Could you please help me to resolve.

[cid:image002.jpg@01D62E85.01FCA170]

File: tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc (hunk 1 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc]: updated based on comment

File: tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc (hunk 3 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc]: updated based on comment

File: tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc (hunk 4 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc]: updated based on comment

File: tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc (hunk 2 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/fully_connected.cc:149]: Please let me know if there is any update

File: tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc (hunk 5 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc:22]: Compared the cycles between Cadence optimization and LUT based solution.
Found that Cadence optimized cycles are better so I suggest to drop LUT based solution.

**Comment by gbaned** [comment]
@bhanuprakashbv It still shows CLA is pending, can you use please make sure to use same GitHub username and email-id associated with it.

**Comment by bhanuprakashbv** [comment]
Hi,

I am using same user name and email-id but I don’t know how to sign corporate CLA.
Could you please help/guide me to do that.

Thanks,
Bhanu

**Comment by bhanuprakashbv** [comment]
@googlebot I signed it!

**Comment by gbaned** [comment]
@bhanuprakashbv  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@bhanuprakashbv Can you please resolve conflicts? Thanks!

**Review by njeffrie — DISMISSED** [review]
Apologies for the delay.  We have shifted optimized kernels around internally to make space for this PR.

**Comment by bhanuprakashbv** [comment]
> Apologies for the delay. We have shifted optimized kernels around internally to make space for this PR.

Hi njeffrie,
Do I need to do any rebase to resolve the conflicting files?
Thanks,
Bhanu

**Comment by rthadur** [comment]
@bhanuprakashbv yes please rebase and resolve conflicts.

**Review by njeffrie — CHANGES_REQUESTED** [review]
It's been brought to my attention that we need a tensorflow license in all source files we check in, similar to what you have in tensorfllow/lite/micro/kernels/xtensa-hifimini/*.cc

Please update all files in this change to make sure the tensorflow license text can be found in all of them.

**Comment by bhanuprakashbv** [comment]
> @bhanuprakashbv yes please rebase and resolve conflicts.
@rthadur did rebase and resolved the conflicts.

**Comment by bhanuprakashbv** [comment]
> It's been brought to my attention that we need a tensorflow license in all source files we check in, similar to what you have in tensorfllow/lite/micro/kernels/xtensa-hifimini/*.cc
> 
> Please update all files in this change to make sure the tensorflow license text can be found in all of them.

@njeffrie Added tensorflow license in all source files. Please check and let me know.
Thanks,
Bhanu

**Review by rthadur — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Review by njeffrie — DISMISSED** [review]
Looks good. We should probably clean up the cadence and TF Licenses throughout the xtensa code to consolidate them into a single TF license. Lets discuss after landing this PR.

File: tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+      err = xa_nn_dot_prod_16x16_asym8s(output_ptr,
+          vector1_ptr,
+          vector2_ptr,
+          bias_ptr,
+          n_memory*n_rank,
+          (data.effective_scale_2_a<<8),
+          data.effective_scale_2_b,
+          output_zp,
+          n_unit);
+      CHECK_ERR_HIFI_NNLIB_KER(err, "xa_nn_dot_prod_16x16_asym8s failed");
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc:200]: `New CHECK or LOG(FATAL) statements inserted in code! Prefer using DCHECK, LOG(DFATAL), or DLOG(FATAL) instead`

File: tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+==============================================================================*/
+
+#ifndef __XTENSA_TF_MICRO_COMMON__
+#define __XTENSA_TF_MICRO_COMMON__
+
+#if defined HIFI_NNLIB_OPT || defined HIFI_MINI_NNLIB_OPT
+#include "xa_nnlib_api.h"
+#include "xa_nnlib_standards.h"
+
+#define CHECK_ERR_HIFI_NNLIB_KER(ret, err_msg) \
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h:44]: `New CHECK or LOG(FATAL) statements inserted in code! Prefer using DCHECK, LOG(DFATAL), or DLOG(FATAL) instead`

File: tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            if(ret != 0) {                     \
+              context->ReportError(            \
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h:46]: `TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`

**Review by rthadur — CHANGES_REQUESTED** [review]
@bhanuprakashbv can you please make below changes , thank you

File: tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+    void* p_scratch = static_cast<void*>(
+        context->GetScratchBuffer(context, op_data->scratch_tensor_index));
+    TFLITE_DCHECK(p_scratch != nullptr);
+
+    for (int i = 0; i < outer_size; ++i) {
+      int err = xa_nn_vec_softmax_asym8s_16(
+          &output_data[i * depth], &input_data[i * depth], op_data->diff_min,
+          op_data->input_left_shift, op_data->input_multiplier, depth, p_scratch);
+      CHECK_ERR_HIFI_NNLIB_KER(err, "xa_nn_vec_softmax_asym8s_16 failed");
```
[the remark this answers is not in the document]
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc]: CHECK_ERR_HIFI_NNLIB_KER is replaced with TF_LITE_KERNEL_LOG

File: tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+                                                     &input[b*n_input],
+                                                     NULL,
+                                                     n_filter,
+                                                     n_input,
+                                                     n_input,
+                                                     n_memory,
+                                                     -input_zp,
+                                                     (data.effective_scale_1_a<<8),
+                                                     data.effective_scale_1_b);
+      CHECK_ERR_HIFI_NNLIB_KER(err, "xa_nn_vec_matXvec_sym8sxasym8s_16 failed");
```
[the remark this answers is not in the document]
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc]: CHECK_ERR_HIFI_NNLIB_KER is replaced with TF_LITE_KERNEL_LOG

File: tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc (hunk 6 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/svdf.cc]: CHECK_ERR_HIFI_NNLIB_KER is replaced with TF_LITE_KERNEL_LOG

File: tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h (hunk 7 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h]: removed CHECK_ERR_HIFI_NNLIB_KER macro definition

File: tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h (hunk 8 above)
Comment (bhanuprakashbv) [reply tensorflow/lite/micro/kernels/xtensa_hifimini/xtensa_tf_micro_common.h]: removed CHECK_ERR_HIFI_NNLIB_KER macro definition

**Comment by bhanuprakashbv** [comment]
> @bhanuprakashbv can you please make below changes , thank you

@rthadur I have updated code based on your comments. Could you please check and let me know.

Thanks,
Bhanu

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.
